File: 백준/Silver/1874. 스택 수열/스택 수열.js
const path = process.platform === 'linux' ? '/dev/stdin' : './input.txt';
let input = require('fs').readFileSync(path).toString().trim().split('\n').map(Number);

const [n, ...sequence] = input;
const stack = [];
let result = '';
let count = 1;

// n번 반복문
for (let i = 0; i < n; i++) {
  while (count <= sequence[i]) {
    stack.push(count);
    result += '+';
    count++;
  }
  let number = stack.pop();
  if (number !== sequence[i]) {
    result = 'NO';
    break;
  }
  result += '-';
}

if (result === 'NO') return console.log('NO');

console.log(result.split('').join('\n'));

File: 백준/Silver/1874. 스택 수열/스택 수열.py
import sys
input = sys.stdin.readline

N = int(input())

stack = []
numbers = [int(input()) for _ in range(N)]
result = []
current = 1
is_impossible = False

for number in numbers:
  if current <= number:
    while current <= number:
      stack.append(current)
      current += 1
      result.append('+')
    stack.pop()
    result.append('-')
  elif stack[-1] == number:
    stack.pop()
    result.append('-')
  else:
    is_impossible = True
    break

if is_impossible:
  print('NO')
else:
  print('\n'.join(result))
